### PR TITLE
Luxor/vested wg spending

### PR DESCRIPTION
**Pull request by dobertRowneySr**
Addresses: #4948 
based on top of #5062 

Note: I have used the extrinsic `vested_transfer` inside the extrinsic `spend_from_budget` all extrinsic in substrate are now `transactional` by default, this means that if the `vested_transfer` fails then `spend_from_budget` is equivalent to a no-op.
Also I have created one internal account for each WG where tokens are minted before they are transferred with vesting.

### CODE REVIEW

**Comment by dobertRowneySr** [comment]
To be rebased onto #5062

**Comment by mnaamani** [comment]
I think there is a simpler way to do this, without needing to have an account on each working group.
Instead of `pay_from_budget` into the working group account then doing `vesting::vested_transfer` you should be able to pay from the budget directly into the target account, then do `vesting::force_vested_transfer` with source and target being the final destination. The balances::transfer inside force_vested_transfer will be a successful no-op, and then the vesting schedule will be applied.

**Comment by mnaamani** [comment]
> Also I have created one internal account for each WG where tokens are minted before they are transferred with vesting.

https://github.com/Joystream/joystream/pull/5072#issuecomment-2002164477

File: runtime-modules/working-group/src/lib.rs [hunk 1]
```diff
@@ -1140,31 +1152,60 @@ decl_module! {
...
+                // we do this to avoid migrations when updating to Luxor
+                let treasury_account_id = T::ModuleId::get().into_sub_account_truncating(I::INDEX);
+                let treasury_amount = Balances::<T>::usable_balance(&treasury_account_id);
+                if treasury_amount < T::ExistentialDeposit::get() {
+                    let treasury_initial_deposit = T::ExistentialDeposit::get().saturating_sub(treasury_amount);
+                    let _ = <balances::Pallet<T>>::deposit_creating(&treasury_account_id, treasury_initial_deposit);
+                }
+
+                // first mint into the treasury account
+                Self::pay_from_budget(&treasury_account_id, amount);
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]: ref: https://github.com/Joystream/joystream/pull/5072#issuecomment-2002164477

```rust
Self::pay_from_budget(account_id, amount);
vesting::Pallet::<T>::force_vested_transfer(
  RawOrigin::Root,
  account_id,
  account_id,
  vesting_schedule
)
```

**Review by mnaamani — CHANGES_REQUESTED** [review]
Can you confirm that [force_vested_transfer](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/vesting/src/lib.rs#L371) the way I'm suggested can be done to simplify this implementation?

**Comment by dobertRowneySr** [comment]
I need to add benchmarks

File: runtime-modules/working-group/src/lib.rs [hunk 2]
```diff
@@ -1626,6 +1681,11 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
             .map(|(worker_id, _)| worker_id)
             .collect()
     }
+
+    /// Returns the account for the  module used for vested spending
+    pub fn module_treasury_account() -> T::AccountId {
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]: this can be dropped

File: runtime-modules/working-group/src/lib.rs [hunk 3]
```diff
@@ -129,6 +137,9 @@ pub trait Config<I: Instance = DefaultInstance>:
 
     /// Stake needed to create an opening
     type LeaderOpeningStake: Get<Self::Balance>;
+
+    /// Type for the Module Id account in order to locked payment the budget.
+    type ModuleId: Get<PalletId>;
```
Comment (mnaamani) [inline runtime-modules/working-group/src/lib.rs]: As we no longer need to store funds in the working group, this can be dropped

File: runtime-modules/working-group/src/tests/fixtures.rs [hunk 4]
```diff
@@ -1159,6 +1162,78 @@ impl SetStatusTextFixture {
...
+
+        let new_budget = TestWorkingGroup::budget();
+        let new_balance = self
+            .vesting_schedule
+            .locked_at::<BlockNumberToBalance>(<frame_system::Pallet<Test>>::block_number());
+        let amount = self.vesting_schedule.locked();
+
+        if actual_result.is_ok() {
+            assert_eq!(new_budget, old_budget - amount);
+            assert_eq!(new_balance, new_balance);
```
Comment (mnaamani) [inline runtime-modules/working-group/src/tests/fixtures.rs]: not sure what is being checked here?

**Review by mnaamani — CHANGES_REQUESTED** [review]

**Comment by mnaamani** [comment]
I think this now needs to be rebased on luxor since I merged the reduce council budget PR

File: runtime-modules/working-group/src/tests/fixtures.rs (hunk 4 above)
Comment (dobertRowneySr) [reply runtime-modules/working-group/src/tests/fixtures.rs]: legacy code that I did not remove..

File: tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts [hunk 5]
```diff
@@ -9,15 +9,27 @@ import { ISubmittableResult } from '@polkadot/types/types/'
...
+export type VestingSchedule = {
+  locked: BN
+  perBlock: BN
+  startingBlock?: BN
+}
+
 export class SpendBudgetFixture extends BaseWorkingGroupFixture {
   protected recievers: string[]
-  protected amounts: BN[]
+  protected vestingSchedules: VestingSchedule[]
```
Comment (thesan) [inline tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts]: Should this be changed ? The fixture is still called `SpendBudgetFixture` not `SpendVestedBudgetFixture`

Comment (dobertRowneySr) [reply tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts]: change reverted, integration tests needs to be taken care of in a separate PR. 
Namely add this Fixture with Flow

File: runtime-modules/content/src/tests/mock.rs [hunk 6]
```diff
@@ -476,6 +479,15 @@ impl common::council::CouncilBudgetManager<U256, u64> for CouncilBudgetManager {
     }
 }
 
+parameter_types! {
+    pub const DWGModuleId: PalletId = PalletId(*b"m_distwg");
+    pub const SWGModuleId: PalletId = PalletId(*b"m_storwg");
+    pub const CWGModuleId: PalletId = PalletId(*b"m_contwg");
```
Comment (mnaamani) [inline runtime-modules/content/src/tests/mock.rs]: ModuleId not required anymore

File: runtime-modules/content/src/tests/mock.rs [hunk 7]
```diff
@@ -490,6 +502,8 @@ impl working_group::Config<StorageWorkingGroupInstance> for Test {
     type WeightInfo = ();
     type MinimumApplicationStake = MinimumApplicationStake;
     type LeaderOpeningStake = LeaderOpeningStake;
+    type VestingBalanceToBalance = BalanceConverter;
+    type ModuleId = SWGModuleId;
```
Comment (mnaamani) [inline runtime-modules/content/src/tests/mock.rs]: ModuleId not a configuration trait anymore so should be removed

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -1355,6 +1355,7 @@ parameter_types! {
     pub const MinimumApplicationStake: Balance = dollars!(20);
     // This should be more costly than `add_opening` fee
     pub const LeaderOpeningStake: Balance = dollars!(100);
+    pub const WorkingGroupModuleId: PalletId = PalletId(*b"mworking"); // module storage
```
Comment (mnaamani) [inline runtime/src/lib.rs:1358]: not required

File: tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts [hunk 9]
```diff
@@ -9,6 +9,12 @@ import { ISubmittableResult } from '@polkadot/types/types/'
 import { Utils } from '../../utils'
 import { BudgetSpendingEventFieldsFragment, WorkingGroupFieldsFragment } from '../../graphql/generated/queries'
 
+export type VestingSchedule = {
```
Comment (mnaamani) [inline tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts:12]: unused type, I think you mentioned integration tests will be updated in a separate PR?

**Review by mnaamani — CHANGES_REQUESTED** [review]
some additional cleanup for build to works after refactor (removing the moduleId in working groups)

File: tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts (hunk 9 above)
Comment (dobertRowneySr) [reply tests/network-tests/src/fixtures/workingGroups/SpendBudgetFixture.ts:12]: yes

**Review by mnaamani — COMMENTED** [review]
Benchmark is failing

```
Error: Input("Benchmark working_group::vested_spend_from_budget failed: AmountLow")
2024-03-29 22:54:25 Starting benchmark: working_group::vested_spend_from_budget    
There was a problem generating the weights for working_group, check the error above
Error: Process completed with exit code 1.
```

File: runtime-modules/working-group/src/benchmarking.rs [hunk 10]
```diff
@@ -787,12 +788,34 @@ benchmarks_instance! {
...
+    }
+
+    vested_spend_from_budget {
+        let (lead_id, _) = insert_a_worker::<T, I>(
+            OpeningType::Leader,
+            0,
+            None
+        );
+
+        let amount_vesting = VestingBalanceOf::<T>::from(10_000u32);
```
Comment (mnaamani) [inline runtime-modules/working-group/src/benchmarking.rs]: Benchmark is failing with error here`AmountLow` because amount is less than vesting pallet configuration of `<T as vesting::Config>::MinVestedTransfer::get()`

Comment (mnaamani) [reply runtime-modules/working-group/src/benchmarking.rs]: So perhaps use that constant to make the `amount_vesting` at least that amount.

File: runtime-modules/working-group/src/benchmarking.rs [hunk 11]
```diff
@@ -787,12 +788,34 @@ benchmarks_instance! {
...
+        let budget = BalanceOf::<T>::from(100_000u32);
+        let block_no = 100u32;
```
Comment (mnaamani) [inline runtime-modules/working-group/src/benchmarking.rs]: `block_no` is not used

**Review by mnaamani — CHANGES_REQUESTED** [review]
from earlier

File: runtime-modules/working-group/src/benchmarking.rs (hunk 11 above)
Comment (ignazio-bovo) [reply runtime-modules/working-group/src/benchmarking.rs:812]: There's a constraint in the runtime lib: 
```rust
parameter_types! {
    pub const MinVestedTransfer: Balance = dollars!(1);
    pub UnvestedFundsAllowedWithdrawReasons: WithdrawReasons = WithdrawReasons::empty();
}
```

I cannot reach 1 dollar with a `u32`, as I was setting `VestedBalance::from(200000000u32)` and it complains if you try to use from with a u64. hence I have added a feature flag to enable this only when runtime benchmarks to disable this during benchmarks.
u32 and u64 are copied by value in rust so there should not be difference in benchmarking result

**Comment by mnaamani** [comment]
strange that I'm seeing errors like `../../query-node/start.sh: line 11: docker-compose: command not found` in CI workflow runs..
Has it been deprecated and we should replace all occurrences of that command with `docker compose` command?

I think you forgot to commit changes to: `types/src/augment/augment-api-rpc.ts` and  `types/src/augment/augment-types.ts`

**Review by mnaamani — APPROVED** [review]
LGTM, next PR to luxor should commit changes to types/src/*.rs missed in this PR

Running integration tests locally on my machine works, we need to investigate why they are failing on github runners.
